In [1]:
import pandas as pd

## Задание 1  
Для датафрейма log из материалов занятия создайте столбец source_type по правилам:

- если источник traffic_source равен Yandex или Google, то в source_type ставится organic;
- для источников paid и email из России ставим ad;
- для источников paid и email не из России ставим other;
- все остальные варианты берём из traffic_source без изменений.

In [2]:
# загружаем файл с логами
log = pd.read_csv('visit_log.csv', sep=';')
log.head()

timestamp    visit_id                                 url  region  \
0  1549980692  e3b0c44298  https://host.ru/3c19b4ef7371864fa3  Russia   
1  1549980704  6e340b9cff  https://host.ru/c8d9213a31839f9a3a  Russia   
2  1549980715  96a296d224  https://host.ru/b8b58337d272ee7b15  Russia   
3  1549980725  709e80c884  https://host.ru/b8b58337d272ee7b15  Russia   
4  1549980736  df3f619804  https://host.ru/b8b58337d272ee7b15  Russia   

      user_id traffic_source  
0  b1613cc09f         yandex  
1  4c3ec14bee         direct  
2  a8c40697fb         yandex  
3  521ac1d6a0         yandex  
4  d7323c571c         yandex

In [3]:
def make_type(row):
    'ф-я для опредления типа источника'
    if row.traffic_source in ['yandex', 'google']:
        return 'organic'
    elif row.traffic_source in ['paid', 'email']:
        if row.region == 'Russia':
            return 'ad'
        else:
            return 'other'
    else:
        return row.traffic_source
    
log['source_type'] = log.apply(make_type, axis=1)

In [4]:
log

timestamp    visit_id                                 url   region  \
0      1549980692  e3b0c44298  https://host.ru/3c19b4ef7371864fa3   Russia   
1      1549980704  6e340b9cff  https://host.ru/c8d9213a31839f9a3a   Russia   
2      1549980715  96a296d224  https://host.ru/b8b58337d272ee7b15   Russia   
3      1549980725  709e80c884  https://host.ru/b8b58337d272ee7b15   Russia   
4      1549980736  df3f619804  https://host.ru/b8b58337d272ee7b15   Russia   
...           ...         ...                                 ...      ...   
18933  1550094288  57e5ba8560  https://host.ru/c2382eb3d6afc8d0f3  Belarus   
18934  1550094296  6f9389ec1b  https://host.ru/f1eb4601740d627ab0   Russia   
18935  1550094308  e8cf2eb8e6  https://host.ru/a5dda93e70318570c0  Belarus   
18936  1550094314  79530b9a67  https://host.ru/6fda01ec57f23abc9e   Russia   
18937  1550094323  b3b634f824  https://host.ru/39fa884393666d45fc   Russia   

          user_id traffic_source source_type  
0      b1613cc09f         yandex     organic  
1      4c3ec14bee         direct      direct  
2      a8c40697fb         yandex     organic  
3      521ac1d6a0         yandex     organic  
4      d7323c571c         yandex     organic  
...           ...            ...         ...  
18933  98b19810d0           paid       other  
18934  32ebb20c13           paid          ad  
18935  b85baa8c73         yandex     organic  
18936  e154b06121           paid          ad  
18937  7d27a58fb8          email          ad  

[18938 rows x 7 columns]

## Задание 2  
В файле URLs.txt содержатся URL страниц новостного сайта. Вам нужно отфильтровать его по адресам страниц с текстами новостей. Известно, что шаблон страницы новостей имеет внутри URL конструкцию: /, затем 8 цифр, затем дефис. Выполните действия:

1) Прочитайте содержимое файла с датафрейм.
2) Отфильтруйте страницы с текстом новостей, используя метод str.contains и регулярное выражение в соответствие с заданным шаблоном.

In [5]:
# считываем файл
urls = pd.read_csv('URLs.txt')
urls.head()

url
0                                            /world/
1                                           /latest/
2                                      /?updated=top
3  /politics/36188461-s-marta-zhizn-rossiyan-susc...
4  /world/36007585-tramp-pridumal-kak-reshit-ukra...

In [6]:
# фильтруем страницы с текстом новостей
urls[urls.url.str.contains('/\d{8}-')]

url
3   /politics/36188461-s-marta-zhizn-rossiyan-susc...
4   /world/36007585-tramp-pridumal-kak-reshit-ukra...
5   /science/36157853-nasa-sobiraet-ekstrennuyu-pr...
6   /video/36001498-poyavilis-pervye-podrobnosti-g...
7   /world/36007585-tramp-pridumal-kak-reshit-ukra...
..                                                ...
88  /cis/35984145-kreml-prokommentiroval-soobschen...
89  /video/36071019-olimpiyskie-obekty-rio-prevrat...
90  /science/36151301-nazvano-posledstvie-zloupotr...
91  /incidents/36027330-vospitatelnitsu-zatravili-...
92  /world/36103095-dominikanskih-zhurnalistov-ubi...

[84 rows x 1 columns]

## Задание 3  
Используйте файл с оценками фильмов ml-latest-small/ratings.csv. Посчитайте среднее время жизни пользователей, которые выставили более 100 оценок. Под временем жизни понимается разница между максимальным и минимальным значениями столбца timestamp для данного значения userId.

In [20]:
# загружаем файл с оценками
rating = pd.read_csv('ratings.csv')
rating.head()

userId  movieId  rating   timestamp
0       1       31     2.5  1260759144
1       1     1029     3.0  1260759179
2       1     1061     3.0  1260759182
3       1     1129     2.0  1260759185
4       1     1172     4.0  1260759205

In [21]:
# посчитаем количество оценок, проставленных каждым userId
count_ratings = rating.groupby('userId').rating.count().reset_index().query('rating > 100')
count_ratings.head()

userId  rating
3        4     204
7        8     116
14      15    1700
16      17     363
18      19     423

In [22]:
# оставим только тех, у кого больше 100 оценок 
rating = rating.merge(count_ratings, on='userId', how='inner', suffixes=('','_count'))
rating.head()

userId  movieId  rating  timestamp  rating_count
0       4       10     4.0  949810645           204
1       4       34     5.0  949919556           204
2       4      112     5.0  949810582           204
3       4      141     5.0  949919681           204
4       4      153     4.0  949811346           204

In [23]:
# для каждого userId найдем время первой и последней оценки
lifetime = rating.groupby('userId').timestamp.min().reset_index().merge(
    rating.groupby('userId').timestamp.max().reset_index(),
    on='userId', how='inner', suffixes=('_min', '_max')
)
lifetime.head()

userId  timestamp_min  timestamp_max
0       4      949778714      949982274
1       8     1154389340     1154474527
2      15      997937239     1469330735
3      17     1127468587     1127476640
4      19      855190091      855195373

In [27]:
# посчитаем среднее время жизни пользователей
# и сразу переведем его из секунд в дни  
(lifetime.timestamp_max - lifetime.timestamp_min).mean() / 86400

463.8947621482917

Cреднее время жизни пользователей - почти 464 дня

## Задание 4  
Дана статистика услуг перевозок клиентов компании по типам.
Нужно сформировать две таблицы:

- таблицу с тремя типами выручки для каждого client_id без указания адреса клиента;
- аналогичную таблицу по типам выручки с указанием адреса клиента.

Дана статистика услуг перевозок клиентов компании по типам:
- rzd - железнодорожные перевозки
- auto - автомобильные перевозки
- air - воздушные перевозки
- client_base - адреса клиентов

In [28]:
rzd = pd.DataFrame(
    {
        'client_id': [111, 112, 113, 114, 115],
        'rzd_revenue': [1093, 2810, 10283, 5774, 981]
    }
)
rzd

client_id  rzd_revenue
0        111         1093
1        112         2810
2        113        10283
3        114         5774
4        115          981

In [29]:
auto = pd.DataFrame(
    {
        'client_id': [113, 114, 115, 116, 117],
        'auto_revenue': [57483, 83, 912, 4834, 98]
    }
)
auto

client_id  auto_revenue
0        113         57483
1        114            83
2        115           912
3        116          4834
4        117            98

In [30]:
air = pd.DataFrame(
    {
        'client_id': [115, 116, 117, 118],
        'air_revenue': [81, 4, 13, 173]
    }
)
air

client_id  air_revenue
0        115           81
1        116            4
2        117           13
3        118          173

In [31]:
client_base = pd.DataFrame(
    {
        'client_id': [111, 112, 113, 114, 115, 116, 117, 118],
        'address': ['Комсомольская 4', 'Энтузиастов 8а', 'Левобережная 1а', 'Мира 14', 'ЗЖБИиДК 1',
                    'Строителей 18', 'Панфиловская 33', 'Мастеркова 4']
    }
)
client_base

client_id          address
0        111  Комсомольская 4
1        112   Энтузиастов 8а
2        113  Левобережная 1а
3        114          Мира 14
4        115        ЗЖБИиДК 1
5        116    Строителей 18
6        117  Панфиловская 33
7        118     Мастеркова 4

In [35]:
# таблица с тремя типами выручки для каждого client_id
client_revenue = rzd.merge(auto, on='client_id', how='outer').merge(
    air, on='client_id', how='outer'
).fillna(0)
client_revenue

client_id  rzd_revenue  auto_revenue  air_revenue
0        111       1093.0           0.0          0.0
1        112       2810.0           0.0          0.0
2        113      10283.0       57483.0          0.0
3        114       5774.0          83.0          0.0
4        115        981.0         912.0         81.0
5        116          0.0        4834.0          4.0
6        117          0.0          98.0         13.0
7        118          0.0           0.0        173.0

In [38]:
# таблица по типам выручки с адресом клиента
client_base_revenue = client_base.merge(client_revenue, on='client_id', how='outer')
client_base_revenue

client_id          address  rzd_revenue  auto_revenue  air_revenue
0        111  Комсомольская 4       1093.0           0.0          0.0
1        112   Энтузиастов 8а       2810.0           0.0          0.0
2        113  Левобережная 1а      10283.0       57483.0          0.0
3        114          Мира 14       5774.0          83.0          0.0
4        115        ЗЖБИиДК 1        981.0         912.0         81.0
5        116    Строителей 18          0.0        4834.0          4.0
6        117  Панфиловская 33          0.0          98.0         13.0
7        118     Мастеркова 4          0.0           0.0        173.0